// mendixmodelsdk/dist/gen/datasets.js
/* tslint:disable */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var internal = require("../sdk/internal");
exports.StructureVersionInfo = internal.StructureVersionInfo;
var projects_1 = require("./projects");
var datasets;
(function (datasets) {
    var DateTimeIntervalLength = (function (_super) {
        __extends(DateTimeIntervalLength, _super);
        function DateTimeIntervalLength() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.qualifiedTsTypeName = "datasets.DateTimeIntervalLength";
            return _this;
        }
        return DateTimeIntervalLength;
    }(internal.IEnum));
    DateTimeIntervalLength.Day = new DateTimeIntervalLength("Day", {});
    DateTimeIntervalLength.Week = new DateTimeIntervalLength("Week", {});
    DateTimeIntervalLength.Period = new DateTimeIntervalLength("Period", {});
    DateTimeIntervalLength.Month = new DateTimeIntervalLength("Month", {});
    DateTimeIntervalLength.Quarter = new DateTimeIntervalLength("Quarter", {});
    DateTimeIntervalLength.Year = new DateTimeIntervalLength("Year", {});
    datasets.DateTimeIntervalLength = DateTimeIntervalLength;
    var DateTimeIntervalModifier = (function (_super) {
        __extends(DateTimeIntervalModifier, _super);
        function DateTimeIntervalModifier() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.qualifiedTsTypeName = "datasets.DateTimeIntervalModifier";
            return _this;
        }
        return DateTimeIntervalModifier;
    }(internal.IEnum));
    DateTimeIntervalModifier.Last = new DateTimeIntervalModifier("Last", {});
    DateTimeIntervalModifier.This = new DateTimeIntervalModifier("This", {});
    DateTimeIntervalModifier.Next = new DateTimeIntervalModifier("Next", {});
    DateTimeIntervalModifier.Past = new DateTimeIntervalModifier("Past", {});
    DateTimeIntervalModifier.Future = new DateTimeIntervalModifier("Future", {});
    DateTimeIntervalModifier.Always = new DateTimeIntervalModifier("Always", {});
    datasets.DateTimeIntervalModifier = DateTimeIntervalModifier;
    /**
     * See: {@link https://docs.mendix.com/refguide7/data-sets relevant section in reference guide}
     */
    var DataSet = (function (_super) {
        __extends(DataSet, _super);
        function DataSet(model, typeName, id, isPartial, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, container) || this;
            _this.__source = new internal.PartProperty(DataSet, _this, "source", null, true);
            _this.__parameters = new internal.PartListProperty(DataSet, _this, "parameters", []);
            _this.__dataSetAccess = new internal.PartProperty(DataSet, _this, "dataSetAccess", null, true);
            _this._containmentName = "documents";
            return _this;
        }
        Object.defineProperty(DataSet.prototype, "containerAsFolderBase", {
            get: function () { return _super.prototype.getContainerAs.call(this, projects_1.projects.FolderBase); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSet.prototype, "source", {
            get: function () { return this.__source.get(); },
            set: function (newValue) { this.__source.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSet.prototype, "parameters", {
            get: function () { return this.__parameters.get(); },
            set: function (newValue) { this.__parameters.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSet.prototype, "dataSetAccess", {
            get: function () { return this.__dataSetAccess.get(); },
            set: function (newValue) { this.__dataSetAccess.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates a new DataSet unit in the SDK and on the server.
         * Expects one argument, the projects.IFolderBase in which this unit is contained.
         */
        DataSet.createIn = function (container) {
            return internal.instancehelpers.createUnit(container, DataSet);
        };
        DataSet.prototype._isByNameReferrable = function () { return true; };
        DataSet.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.dataSetAccess = DataSetAccess.create(this.model);
            this.source = OqlDataSetSource.create(this.model);
        };
        return DataSet;
    }(projects_1.projects.Document));
    DataSet.typeName = "DataSets$DataSet";
    DataSet.versionInfo = new exports.StructureVersionInfo({
        "properties": {
            "source": {
                "required": {
                    "currentValue": true
                }
            },
            "parameters": {
                "public": {
                    "currentValue": true
                }
            },
            "dataSetAccess": {
                "required": {
                    "currentValue": true
                }
            }
        }
    }, internal.StructureType.ModelUnit);
    datasets.DataSet = DataSet;
    var DataSetAccess = (function (_super) {
        __extends(DataSetAccess, _super);
        function DataSetAccess(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__moduleRoleAccessList = new internal.PartListProperty(DataSetAccess, _this, "moduleRoleAccessList", []);
            if (arguments.length < 4) {
                throw new Error("new DataSetAccess() cannot be invoked directly, please use 'model.datasets.createDataSetAccess()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetAccess.prototype, "containerAsDataSet", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSet); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetAccess.prototype, "moduleRoleAccessList", {
            get: function () { return this.__moduleRoleAccessList.get(); },
            set: function (newValue) { this.__moduleRoleAccessList.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetAccess instance in the SDK and on the server.
         * The new DataSetAccess will be automatically stored in the 'dataSetAccess' property
         * of the parent DataSet element passed as argument.
         */
        DataSetAccess.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetAccess, "dataSetAccess", false);
        };
        /**
         * Creates and returns a new DataSetAccess instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetAccess.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetAccess);
        };
        DataSetAccess.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DataSetAccess;
    }(internal.Element));
    DataSetAccess.typeName = "DataSets$DataSetAccess";
    DataSetAccess.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetAccess = DataSetAccess;
    var DataSetColumn = (function (_super) {
        __extends(DataSetColumn, _super);
        function DataSetColumn(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__name = new internal.PrimitiveProperty(DataSetColumn, _this, "name", "", internal.PrimitiveTypeEnum.String);
            _this.__type = new internal.PrimitiveProperty(DataSetColumn, _this, "type", "", internal.PrimitiveTypeEnum.String);
            if (arguments.length < 4) {
                throw new Error("new DataSetColumn() cannot be invoked directly, please use 'model.datasets.createDataSetColumn()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetColumn.prototype, "containerAsJavaDataSetSource", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaDataSetSource); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetColumn.prototype, "name", {
            get: function () { return this.__name.get(); },
            set: function (newValue) { this.__name.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetColumn.prototype, "type", {
            /**
             * The value of this property is conceptually of type dataTypes.DataType.
             */
            get: function () { return this.__type.get(); },
            set: function (newValue) { this.__type.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetColumn instance in the SDK and on the server.
         * The new DataSetColumn will be automatically stored in the 'columns' property
         * of the parent JavaDataSetSource element passed as argument.
         */
        DataSetColumn.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetColumn, "columns", true);
        };
        /**
         * Creates and returns a new DataSetColumn instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetColumn.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetColumn);
        };
        DataSetColumn.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.name = "Column";
            this.type = "String";
        };
        return DataSetColumn;
    }(internal.Element));
    DataSetColumn.typeName = "DataSets$DataSetColumn";
    DataSetColumn.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetColumn = DataSetColumn;
    var DataSetConstraintAccess = (function (_super) {
        __extends(DataSetConstraintAccess, _super);
        function DataSetConstraintAccess(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__constraintText = new internal.PrimitiveProperty(DataSetConstraintAccess, _this, "constraintText", "", internal.PrimitiveTypeEnum.String);
            _this.__enabled = new internal.PrimitiveProperty(DataSetConstraintAccess, _this, "enabled", false, internal.PrimitiveTypeEnum.Boolean);
            if (arguments.length < 4) {
                throw new Error("new DataSetConstraintAccess() cannot be invoked directly, please use 'model.datasets.createDataSetConstraintAccess()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetConstraintAccess.prototype, "containerAsDataSetParameterAccess", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSetParameterAccess); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetConstraintAccess.prototype, "constraintText", {
            get: function () { return this.__constraintText.get(); },
            set: function (newValue) { this.__constraintText.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetConstraintAccess.prototype, "enabled", {
            get: function () { return this.__enabled.get(); },
            set: function (newValue) { this.__enabled.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetConstraintAccess instance in the SDK and on the server.
         * The new DataSetConstraintAccess will be automatically stored in the 'constraintAccessList' property
         * of the parent DataSetParameterAccess element passed as argument.
         */
        DataSetConstraintAccess.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetConstraintAccess, "constraintAccessList", true);
        };
        /**
         * Creates and returns a new DataSetConstraintAccess instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetConstraintAccess.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetConstraintAccess);
        };
        DataSetConstraintAccess.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DataSetConstraintAccess;
    }(internal.Element));
    DataSetConstraintAccess.typeName = "DataSets$DataSetConstraintAccess";
    DataSetConstraintAccess.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetConstraintAccess = DataSetConstraintAccess;
    /**
     * Warning: this class represents an abstract and non-instantiable concept in the meta model.
     * You should _not_ try to instantiate this class directly, but instantiate one of its concrete/non-abstract sub classes instead.
     * Instantiating this class is possible in pure JavaScript, but there's no guarantee the result will be at all useful.
     */
    var DataSetParameterConstraint = (function (_super) {
        __extends(DataSetParameterConstraint, _super);
        function DataSetParameterConstraint(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new DataSetParameterConstraint() cannot be invoked directly, please use 'model.datasets.createDataSetParameterConstraint()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetParameterConstraint.prototype, "containerAsDataSetParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSetParameter); },
            enumerable: true,
            configurable: true
        });
        DataSetParameterConstraint.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DataSetParameterConstraint;
    }(internal.Element));
    DataSetParameterConstraint.typeName = "DataSets$DataSetParameterConstraint";
    DataSetParameterConstraint.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetParameterConstraint = DataSetParameterConstraint;
    var DataSetDateTimeConstraint = (function (_super) {
        __extends(DataSetDateTimeConstraint, _super);
        function DataSetDateTimeConstraint(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__modifier = new internal.EnumProperty(DataSetDateTimeConstraint, _this, "modifier", DateTimeIntervalModifier.Last, DateTimeIntervalModifier);
            _this.__length = new internal.EnumProperty(DataSetDateTimeConstraint, _this, "length", DateTimeIntervalLength.Month, DateTimeIntervalLength);
            if (arguments.length < 4) {
                throw new Error("new DataSetDateTimeConstraint() cannot be invoked directly, please use 'model.datasets.createDataSetDateTimeConstraint()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetDateTimeConstraint.prototype, "containerAsDataSetParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSetParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetDateTimeConstraint.prototype, "modifier", {
            get: function () { return this.__modifier.get(); },
            set: function (newValue) { this.__modifier.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetDateTimeConstraint.prototype, "length", {
            get: function () { return this.__length.get(); },
            set: function (newValue) { this.__length.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetDateTimeConstraint instance in the SDK and on the server.
         * The new DataSetDateTimeConstraint will be automatically stored in the 'constraints' property
         * of the parent DataSetParameter element passed as argument.
         */
        DataSetDateTimeConstraint.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetDateTimeConstraint, "constraints", true);
        };
        /**
         * Creates and returns a new DataSetDateTimeConstraint instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetDateTimeConstraint.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetDateTimeConstraint);
        };
        DataSetDateTimeConstraint.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.length = DateTimeIntervalLength.Month;
            this.modifier = DateTimeIntervalModifier.Last;
        };
        return DataSetDateTimeConstraint;
    }(DataSetParameterConstraint));
    DataSetDateTimeConstraint.typeName = "DataSets$DataSetDateTimeConstraint";
    DataSetDateTimeConstraint.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetDateTimeConstraint = DataSetDateTimeConstraint;
    var DataSetModuleRoleAccess = (function (_super) {
        __extends(DataSetModuleRoleAccess, _super);
        function DataSetModuleRoleAccess(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__parameterAccessList = new internal.PartListProperty(DataSetModuleRoleAccess, _this, "parameterAccessList", []);
            _this.__moduleRole = new internal.ByNameReferenceProperty(DataSetModuleRoleAccess, _this, "moduleRole", null, "Security$ModuleRole");
            if (arguments.length < 4) {
                throw new Error("new DataSetModuleRoleAccess() cannot be invoked directly, please use 'model.datasets.createDataSetModuleRoleAccess()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetModuleRoleAccess.prototype, "containerAsDataSetAccess", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSetAccess); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetModuleRoleAccess.prototype, "parameterAccessList", {
            get: function () { return this.__parameterAccessList.get(); },
            set: function (newValue) { this.__parameterAccessList.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetModuleRoleAccess.prototype, "moduleRole", {
            get: function () { return this.__moduleRole.get(); },
            set: function (newValue) { this.__moduleRole.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetModuleRoleAccess.prototype, "moduleRoleQualifiedName", {
            get: function () { return this.__moduleRole.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetModuleRoleAccess instance in the SDK and on the server.
         * The new DataSetModuleRoleAccess will be automatically stored in the 'moduleRoleAccessList' property
         * of the parent DataSetAccess element passed as argument.
         */
        DataSetModuleRoleAccess.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetModuleRoleAccess, "moduleRoleAccessList", true);
        };
        /**
         * Creates and returns a new DataSetModuleRoleAccess instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetModuleRoleAccess.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetModuleRoleAccess);
        };
        DataSetModuleRoleAccess.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DataSetModuleRoleAccess;
    }(internal.Element));
    DataSetModuleRoleAccess.typeName = "DataSets$DataSetModuleRoleAccess";
    DataSetModuleRoleAccess.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetModuleRoleAccess = DataSetModuleRoleAccess;
    var DataSetNumericConstraint = (function (_super) {
        __extends(DataSetNumericConstraint, _super);
        function DataSetNumericConstraint(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__begin = new internal.PrimitiveProperty(DataSetNumericConstraint, _this, "begin", "", internal.PrimitiveTypeEnum.String);
            _this.__applyBegin = new internal.PrimitiveProperty(DataSetNumericConstraint, _this, "applyBegin", false, internal.PrimitiveTypeEnum.Boolean);
            _this.__end = new internal.PrimitiveProperty(DataSetNumericConstraint, _this, "end", "", internal.PrimitiveTypeEnum.String);
            _this.__applyEnd = new internal.PrimitiveProperty(DataSetNumericConstraint, _this, "applyEnd", false, internal.PrimitiveTypeEnum.Boolean);
            if (arguments.length < 4) {
                throw new Error("new DataSetNumericConstraint() cannot be invoked directly, please use 'model.datasets.createDataSetNumericConstraint()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetNumericConstraint.prototype, "containerAsDataSetParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSetParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetNumericConstraint.prototype, "begin", {
            get: function () { return this.__begin.get(); },
            set: function (newValue) { this.__begin.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetNumericConstraint.prototype, "applyBegin", {
            get: function () { return this.__applyBegin.get(); },
            set: function (newValue) { this.__applyBegin.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetNumericConstraint.prototype, "end", {
            get: function () { return this.__end.get(); },
            set: function (newValue) { this.__end.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetNumericConstraint.prototype, "applyEnd", {
            get: function () { return this.__applyEnd.get(); },
            set: function (newValue) { this.__applyEnd.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetNumericConstraint instance in the SDK and on the server.
         * The new DataSetNumericConstraint will be automatically stored in the 'constraints' property
         * of the parent DataSetParameter element passed as argument.
         */
        DataSetNumericConstraint.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetNumericConstraint, "constraints", true);
        };
        /**
         * Creates and returns a new DataSetNumericConstraint instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetNumericConstraint.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetNumericConstraint);
        };
        DataSetNumericConstraint.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.applyBegin = true;
            this.applyEnd = true;
            this.begin = "0";
            this.end = "0";
        };
        return DataSetNumericConstraint;
    }(DataSetParameterConstraint));
    DataSetNumericConstraint.typeName = "DataSets$DataSetNumericConstraint";
    DataSetNumericConstraint.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetNumericConstraint = DataSetNumericConstraint;
    var DataSetObjectConstraint = (function (_super) {
        __extends(DataSetObjectConstraint, _super);
        function DataSetObjectConstraint(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__name = new internal.PrimitiveProperty(DataSetObjectConstraint, _this, "name", "", internal.PrimitiveTypeEnum.String);
            _this.__constraint = new internal.PrimitiveProperty(DataSetObjectConstraint, _this, "constraint", "", internal.PrimitiveTypeEnum.String);
            if (arguments.length < 4) {
                throw new Error("new DataSetObjectConstraint() cannot be invoked directly, please use 'model.datasets.createDataSetObjectConstraint()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetObjectConstraint.prototype, "containerAsDataSetParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSetParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetObjectConstraint.prototype, "name", {
            get: function () { return this.__name.get(); },
            set: function (newValue) { this.__name.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetObjectConstraint.prototype, "constraint", {
            /**
             * The value of this property is conceptually of type xPathConstraints.XPathConstraint.
             */
            get: function () { return this.__constraint.get(); },
            set: function (newValue) { this.__constraint.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetObjectConstraint instance in the SDK and on the server.
         * The new DataSetObjectConstraint will be automatically stored in the 'constraints' property
         * of the parent DataSetParameter element passed as argument.
         */
        DataSetObjectConstraint.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetObjectConstraint, "constraints", true);
        };
        /**
         * Creates and returns a new DataSetObjectConstraint instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetObjectConstraint.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetObjectConstraint);
        };
        DataSetObjectConstraint.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DataSetObjectConstraint;
    }(DataSetParameterConstraint));
    DataSetObjectConstraint.typeName = "DataSets$DataSetObjectConstraint";
    DataSetObjectConstraint.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetObjectConstraint = DataSetObjectConstraint;
    var DataSetParameter = (function (_super) {
        __extends(DataSetParameter, _super);
        function DataSetParameter(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__name = new internal.PrimitiveProperty(DataSetParameter, _this, "name", "", internal.PrimitiveTypeEnum.String);
            _this.__type = new internal.PrimitiveProperty(DataSetParameter, _this, "type", "", internal.PrimitiveTypeEnum.String);
            _this.__constraints = new internal.PartListProperty(DataSetParameter, _this, "constraints", []);
            if (arguments.length < 4) {
                throw new Error("new DataSetParameter() cannot be invoked directly, please use 'model.datasets.createDataSetParameter()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetParameter.prototype, "containerAsDataSet", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSet); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetParameter.prototype, "name", {
            get: function () { return this.__name.get(); },
            set: function (newValue) { this.__name.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetParameter.prototype, "type", {
            /**
             * The value of this property is conceptually of type dataTypes.DataType.
             */
            get: function () { return this.__type.get(); },
            set: function (newValue) { this.__type.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetParameter.prototype, "constraints", {
            get: function () { return this.__constraints.get(); },
            set: function (newValue) { this.__constraints.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetParameter instance in the SDK and on the server.
         * The new DataSetParameter will be automatically stored in the 'parameters' property
         * of the parent DataSet element passed as argument.
         */
        DataSetParameter.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetParameter, "parameters", true);
        };
        /**
         * Creates and returns a new DataSetParameter instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetParameter.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetParameter);
        };
        DataSetParameter.prototype._isByNameReferrable = function () { return true; };
        DataSetParameter.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.type = "Unknown";
        };
        return DataSetParameter;
    }(internal.Element));
    DataSetParameter.typeName = "DataSets$DataSetParameter";
    DataSetParameter.versionInfo = new exports.StructureVersionInfo({
        "properties": {
            "name": {
                "public": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    datasets.DataSetParameter = DataSetParameter;
    var DataSetParameterAccess = (function (_super) {
        __extends(DataSetParameterAccess, _super);
        function DataSetParameterAccess(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__parameterName = new internal.PrimitiveProperty(DataSetParameterAccess, _this, "parameterName", "", internal.PrimitiveTypeEnum.String);
            _this.__constraintAccessList = new internal.PartListProperty(DataSetParameterAccess, _this, "constraintAccessList", []);
            if (arguments.length < 4) {
                throw new Error("new DataSetParameterAccess() cannot be invoked directly, please use 'model.datasets.createDataSetParameterAccess()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetParameterAccess.prototype, "containerAsDataSetModuleRoleAccess", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSetModuleRoleAccess); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetParameterAccess.prototype, "parameterName", {
            get: function () { return this.__parameterName.get(); },
            set: function (newValue) { this.__parameterName.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DataSetParameterAccess.prototype, "constraintAccessList", {
            get: function () { return this.__constraintAccessList.get(); },
            set: function (newValue) { this.__constraintAccessList.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DataSetParameterAccess instance in the SDK and on the server.
         * The new DataSetParameterAccess will be automatically stored in the 'parameterAccessList' property
         * of the parent DataSetModuleRoleAccess element passed as argument.
         */
        DataSetParameterAccess.createIn = function (container) {
            return internal.instancehelpers.createElement(container, DataSetParameterAccess, "parameterAccessList", true);
        };
        /**
         * Creates and returns a new DataSetParameterAccess instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DataSetParameterAccess.create = function (model) {
            return internal.instancehelpers.createElement(model, DataSetParameterAccess);
        };
        DataSetParameterAccess.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DataSetParameterAccess;
    }(internal.Element));
    DataSetParameterAccess.typeName = "DataSets$DataSetParameterAccess";
    DataSetParameterAccess.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetParameterAccess = DataSetParameterAccess;
    /**
     * Warning: this class represents an abstract and non-instantiable concept in the meta model.
     * You should _not_ try to instantiate this class directly, but instantiate one of its concrete/non-abstract sub classes instead.
     * Instantiating this class is possible in pure JavaScript, but there's no guarantee the result will be at all useful.
     */
    var DataSetSource = (function (_super) {
        __extends(DataSetSource, _super);
        function DataSetSource(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new DataSetSource() cannot be invoked directly, please use 'model.datasets.createDataSetSource()'");
            }
            return _this;
        }
        Object.defineProperty(DataSetSource.prototype, "containerAsDataSet", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSet); },
            enumerable: true,
            configurable: true
        });
        DataSetSource.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DataSetSource;
    }(internal.Element));
    DataSetSource.typeName = "DataSets$DataSetSource";
    DataSetSource.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.DataSetSource = DataSetSource;
    var JavaDataSetSource = (function (_super) {
        __extends(JavaDataSetSource, _super);
        function JavaDataSetSource(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__columns = new internal.PartListProperty(JavaDataSetSource, _this, "columns", []);
            if (arguments.length < 4) {
                throw new Error("new JavaDataSetSource() cannot be invoked directly, please use 'model.datasets.createJavaDataSetSource()'");
            }
            return _this;
        }
        Object.defineProperty(JavaDataSetSource.prototype, "containerAsDataSet", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSet); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaDataSetSource.prototype, "columns", {
            get: function () { return this.__columns.get(); },
            set: function (newValue) { this.__columns.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new JavaDataSetSource instance in the SDK and on the server.
         * The new JavaDataSetSource will be automatically stored in the 'source' property
         * of the parent DataSet element passed as argument.
         */
        JavaDataSetSource.createIn = function (container) {
            return internal.instancehelpers.createElement(container, JavaDataSetSource, "source", false);
        };
        /**
         * Creates and returns a new JavaDataSetSource instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        JavaDataSetSource.create = function (model) {
            return internal.instancehelpers.createElement(model, JavaDataSetSource);
        };
        JavaDataSetSource.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return JavaDataSetSource;
    }(DataSetSource));
    JavaDataSetSource.typeName = "DataSets$JavaDataSetSource";
    JavaDataSetSource.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.JavaDataSetSource = JavaDataSetSource;
    var OqlDataSetSource = (function (_super) {
        __extends(OqlDataSetSource, _super);
        function OqlDataSetSource(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__query = new internal.PrimitiveProperty(OqlDataSetSource, _this, "query", "", internal.PrimitiveTypeEnum.String);
            _this.__ignoreErrorsInQuery = new internal.PrimitiveProperty(OqlDataSetSource, _this, "ignoreErrorsInQuery", false, internal.PrimitiveTypeEnum.Boolean);
            if (arguments.length < 4) {
                throw new Error("new OqlDataSetSource() cannot be invoked directly, please use 'model.datasets.createOqlDataSetSource()'");
            }
            return _this;
        }
        Object.defineProperty(OqlDataSetSource.prototype, "containerAsDataSet", {
            get: function () { return _super.prototype.getContainerAs.call(this, DataSet); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(OqlDataSetSource.prototype, "query", {
            /**
             * The value of this property is conceptually of type oql.OqlQuery.
             */
            get: function () { return this.__query.get(); },
            set: function (newValue) { this.__query.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(OqlDataSetSource.prototype, "ignoreErrorsInQuery", {
            get: function () { return this.__ignoreErrorsInQuery.get(); },
            set: function (newValue) { this.__ignoreErrorsInQuery.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new OqlDataSetSource instance in the SDK and on the server.
         * The new OqlDataSetSource will be automatically stored in the 'source' property
         * of the parent DataSet element passed as argument.
         */
        OqlDataSetSource.createIn = function (container) {
            return internal.instancehelpers.createElement(container, OqlDataSetSource, "source", false);
        };
        /**
         * Creates and returns a new OqlDataSetSource instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        OqlDataSetSource.create = function (model) {
            return internal.instancehelpers.createElement(model, OqlDataSetSource);
        };
        OqlDataSetSource.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return OqlDataSetSource;
    }(DataSetSource));
    OqlDataSetSource.typeName = "DataSets$OqlDataSetSource";
    OqlDataSetSource.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    datasets.OqlDataSetSource = OqlDataSetSource;
})(datasets = exports.datasets || (exports.datasets = {}));
