// mendixmodelsdk/dist/gen/javaactions.js
/* tslint:disable */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var internal = require("../sdk/internal");
exports.StructureVersionInfo = internal.StructureVersionInfo;
var projects_1 = require("./projects");
var javaactions;
(function (javaactions) {
    /**
     * Interfaces and instance classes for types from the Mendix sub meta model `JavaActions`.
     */
    /**
     * Warning: this class represents an abstract and non-instantiable concept in the meta model.
     * You should _not_ try to instantiate this class directly, but instantiate one of its concrete/non-abstract sub classes instead.
     * Instantiating this class is possible in pure JavaScript, but there's no guarantee the result will be at all useful.

     *
     * In version 6.7.0: introduced
     */
    var ParameterType = (function (_super) {
        __extends(ParameterType, _super);
        function ParameterType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new ParameterType() cannot be invoked directly, please use 'model.javaactions.createParameterType()'");
            }
            return _this;
        }
        Object.defineProperty(ParameterType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        ParameterType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return ParameterType;
    }(internal.Element));
    ParameterType.typeName = "JavaActions$ParameterType";
    ParameterType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.7.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.ParameterType = ParameterType;
    /**
     * In version 6.7.0: introduced
     */
    var BasicParameterType = (function (_super) {
        __extends(BasicParameterType, _super);
        function BasicParameterType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__type = new internal.PartProperty(BasicParameterType, _this, "type", null, true);
            if (arguments.length < 4) {
                throw new Error("new BasicParameterType() cannot be invoked directly, please use 'model.javaactions.createBasicParameterType()'");
            }
            return _this;
        }
        Object.defineProperty(BasicParameterType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(BasicParameterType.prototype, "type", {
            get: function () { return this.__type.get(); },
            set: function (newValue) { this.__type.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new BasicParameterType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        BasicParameterType.create = function (model) {
            return internal.instancehelpers.createElement(model, BasicParameterType);
        };
        BasicParameterType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.type = StringType.create(this.model);
        };
        return BasicParameterType;
    }(ParameterType));
    BasicParameterType.typeName = "JavaActions$BasicParameterType";
    BasicParameterType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.7.0",
        "properties": {
            "type": {
                "public": {
                    "currentValue": true
                },
                "required": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.BasicParameterType = BasicParameterType;
    /**
     * Warning: this class represents an abstract and non-instantiable concept in the meta model.
     * You should _not_ try to instantiate this class directly, but instantiate one of its concrete/non-abstract sub classes instead.
     * Instantiating this class is possible in pure JavaScript, but there's no guarantee the result will be at all useful.

     *
     * In version 6.6.0: introduced
     */
    var Type = (function (_super) {
        __extends(Type, _super);
        function Type(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new Type() cannot be invoked directly, please use 'model.javaactions.createType()'");
            }
            return _this;
        }
        Object.defineProperty(Type.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Type.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Type.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(Type.prototype, "containerAsListType", {
            get: function () { return _super.prototype.getContainerAs.call(this, ListType); },
            enumerable: true,
            configurable: true
        });
        Type.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return Type;
    }(internal.Element));
    Type.typeName = "JavaActions$Type";
    Type.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.Type = Type;
    /**
     * Warning: this class represents an abstract and non-instantiable concept in the meta model.
     * You should _not_ try to instantiate this class directly, but instantiate one of its concrete/non-abstract sub classes instead.
     * Instantiating this class is possible in pure JavaScript, but there's no guarantee the result will be at all useful.

     *
     * In version 6.6.0: introduced
     */
    var PrimitiveType = (function (_super) {
        __extends(PrimitiveType, _super);
        function PrimitiveType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new PrimitiveType() cannot be invoked directly, please use 'model.javaactions.createPrimitiveType()'");
            }
            return _this;
        }
        Object.defineProperty(PrimitiveType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PrimitiveType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PrimitiveType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        PrimitiveType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return PrimitiveType;
    }(Type));
    PrimitiveType.typeName = "JavaActions$PrimitiveType";
    PrimitiveType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.PrimitiveType = PrimitiveType;
    /**
     * In version 6.6.0: introduced
     */
    var BooleanType = (function (_super) {
        __extends(BooleanType, _super);
        function BooleanType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new BooleanType() cannot be invoked directly, please use 'model.javaactions.createBooleanType()'");
            }
            return _this;
        }
        Object.defineProperty(BooleanType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(BooleanType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(BooleanType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new BooleanType instance in the SDK and on the server.
         * The new BooleanType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        BooleanType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, BooleanType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, BooleanType, "type", false);
        };
        /**
         * Creates and returns a new BooleanType instance in the SDK and on the server.
         * The new BooleanType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        BooleanType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, BooleanType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, BooleanType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new BooleanType instance in the SDK and on the server.
         * The new BooleanType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        BooleanType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, BooleanType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, BooleanType, "javaType", false);
        };
        /**
         * Creates and returns a new BooleanType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        BooleanType.create = function (model) {
            return internal.instancehelpers.createElement(model, BooleanType);
        };
        BooleanType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return BooleanType;
    }(PrimitiveType));
    BooleanType.typeName = "JavaActions$BooleanType";
    BooleanType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.BooleanType = BooleanType;
    /**
     * Warning: this class represents an abstract and non-instantiable concept in the meta model.
     * You should _not_ try to instantiate this class directly, but instantiate one of its concrete/non-abstract sub classes instead.
     * Instantiating this class is possible in pure JavaScript, but there's no guarantee the result will be at all useful.

     *
     * In version 6.6.0: introduced
     */
    var EntityType = (function (_super) {
        __extends(EntityType, _super);
        function EntityType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new EntityType() cannot be invoked directly, please use 'model.javaactions.createEntityType()'");
            }
            return _this;
        }
        Object.defineProperty(EntityType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EntityType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EntityType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EntityType.prototype, "containerAsListType", {
            get: function () { return _super.prototype.getContainerAs.call(this, ListType); },
            enumerable: true,
            configurable: true
        });
        EntityType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return EntityType;
    }(Type));
    EntityType.typeName = "JavaActions$EntityType";
    EntityType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.EntityType = EntityType;
    /**
     * In version 6.6.0: introduced
     */
    var ConcreteEntityType = (function (_super) {
        __extends(ConcreteEntityType, _super);
        function ConcreteEntityType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__entity = new internal.ByNameReferenceProperty(ConcreteEntityType, _this, "entity", null, "DomainModels$Entity");
            if (arguments.length < 4) {
                throw new Error("new ConcreteEntityType() cannot be invoked directly, please use 'model.javaactions.createConcreteEntityType()'");
            }
            return _this;
        }
        Object.defineProperty(ConcreteEntityType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ConcreteEntityType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ConcreteEntityType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ConcreteEntityType.prototype, "containerAsListType", {
            get: function () { return _super.prototype.getContainerAs.call(this, ListType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ConcreteEntityType.prototype, "entity", {
            get: function () { return this.__entity.get(); },
            set: function (newValue) { this.__entity.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ConcreteEntityType.prototype, "entityQualifiedName", {
            get: function () { return this.__entity.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new ConcreteEntityType instance in the SDK and on the server.
         * The new ConcreteEntityType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        ConcreteEntityType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, ConcreteEntityType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, ConcreteEntityType, "type", false);
        };
        /**
         * Creates and returns a new ConcreteEntityType instance in the SDK and on the server.
         * The new ConcreteEntityType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        ConcreteEntityType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, ConcreteEntityType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, ConcreteEntityType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new ConcreteEntityType instance in the SDK and on the server.
         * The new ConcreteEntityType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        ConcreteEntityType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, ConcreteEntityType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, ConcreteEntityType, "javaType", false);
        };
        /**
         * Creates and returns a new ConcreteEntityType instance in the SDK and on the server.
         * The new ConcreteEntityType will be automatically stored in the 'parameter' property
         * of the parent ListType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        ConcreteEntityType.createInListTypeUnderParameter = function (container) {
            internal.createInVersionCheck(container.model, ConcreteEntityType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, ConcreteEntityType, "parameter", false);
        };
        /**
         * Creates and returns a new ConcreteEntityType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        ConcreteEntityType.create = function (model) {
            return internal.instancehelpers.createElement(model, ConcreteEntityType);
        };
        ConcreteEntityType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return ConcreteEntityType;
    }(EntityType));
    ConcreteEntityType.typeName = "JavaActions$ConcreteEntityType";
    ConcreteEntityType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "properties": {
            "entity": {
                "public": {
                    "currentValue": true
                },
                "required": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.ConcreteEntityType = ConcreteEntityType;
    /**
     * In version 6.6.0: introduced
     */
    var DateTimeType = (function (_super) {
        __extends(DateTimeType, _super);
        function DateTimeType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new DateTimeType() cannot be invoked directly, please use 'model.javaactions.createDateTimeType()'");
            }
            return _this;
        }
        Object.defineProperty(DateTimeType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DateTimeType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DateTimeType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DateTimeType instance in the SDK and on the server.
         * The new DateTimeType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        DateTimeType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, DateTimeType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, DateTimeType, "type", false);
        };
        /**
         * Creates and returns a new DateTimeType instance in the SDK and on the server.
         * The new DateTimeType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        DateTimeType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, DateTimeType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, DateTimeType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new DateTimeType instance in the SDK and on the server.
         * The new DateTimeType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        DateTimeType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, DateTimeType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, DateTimeType, "javaType", false);
        };
        /**
         * Creates and returns a new DateTimeType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DateTimeType.create = function (model) {
            return internal.instancehelpers.createElement(model, DateTimeType);
        };
        DateTimeType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DateTimeType;
    }(PrimitiveType));
    DateTimeType.typeName = "JavaActions$DateTimeType";
    DateTimeType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.DateTimeType = DateTimeType;
    /**
     * In version 6.6.0: introduced
     */
    var DecimalType = (function (_super) {
        __extends(DecimalType, _super);
        function DecimalType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new DecimalType() cannot be invoked directly, please use 'model.javaactions.createDecimalType()'");
            }
            return _this;
        }
        Object.defineProperty(DecimalType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DecimalType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(DecimalType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new DecimalType instance in the SDK and on the server.
         * The new DecimalType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        DecimalType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, DecimalType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, DecimalType, "type", false);
        };
        /**
         * Creates and returns a new DecimalType instance in the SDK and on the server.
         * The new DecimalType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        DecimalType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, DecimalType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, DecimalType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new DecimalType instance in the SDK and on the server.
         * The new DecimalType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        DecimalType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, DecimalType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, DecimalType, "javaType", false);
        };
        /**
         * Creates and returns a new DecimalType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        DecimalType.create = function (model) {
            return internal.instancehelpers.createElement(model, DecimalType);
        };
        DecimalType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return DecimalType;
    }(PrimitiveType));
    DecimalType.typeName = "JavaActions$DecimalType";
    DecimalType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.DecimalType = DecimalType;
    /**
     * In version 6.7.0: introduced
     */
    var EntityTypeParameterType = (function (_super) {
        __extends(EntityTypeParameterType, _super);
        function EntityTypeParameterType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__typeParameter = new internal.ByIdReferenceProperty(EntityTypeParameterType, _this, "typeParameter", null);
            if (arguments.length < 4) {
                throw new Error("new EntityTypeParameterType() cannot be invoked directly, please use 'model.javaactions.createEntityTypeParameterType()'");
            }
            return _this;
        }
        Object.defineProperty(EntityTypeParameterType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EntityTypeParameterType.prototype, "typeParameter", {
            get: function () { return this.__typeParameter.get(); },
            set: function (newValue) { this.__typeParameter.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new EntityTypeParameterType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        EntityTypeParameterType.create = function (model) {
            return internal.instancehelpers.createElement(model, EntityTypeParameterType);
        };
        EntityTypeParameterType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return EntityTypeParameterType;
    }(ParameterType));
    EntityTypeParameterType.typeName = "JavaActions$EntityTypeParameterType";
    EntityTypeParameterType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.7.0",
        "properties": {
            "typeParameter": {
                "public": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.EntityTypeParameterType = EntityTypeParameterType;
    /**
     * In version 6.6.0: introduced
     */
    var EnumerationType = (function (_super) {
        __extends(EnumerationType, _super);
        function EnumerationType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__enumeration = new internal.ByNameReferenceProperty(EnumerationType, _this, "enumeration", null, "Enumerations$Enumeration");
            if (arguments.length < 4) {
                throw new Error("new EnumerationType() cannot be invoked directly, please use 'model.javaactions.createEnumerationType()'");
            }
            return _this;
        }
        Object.defineProperty(EnumerationType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EnumerationType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EnumerationType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EnumerationType.prototype, "enumeration", {
            get: function () { return this.__enumeration.get(); },
            set: function (newValue) { this.__enumeration.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EnumerationType.prototype, "enumerationQualifiedName", {
            get: function () { return this.__enumeration.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new EnumerationType instance in the SDK and on the server.
         * The new EnumerationType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        EnumerationType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, EnumerationType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, EnumerationType, "type", false);
        };
        /**
         * Creates and returns a new EnumerationType instance in the SDK and on the server.
         * The new EnumerationType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        EnumerationType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, EnumerationType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, EnumerationType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new EnumerationType instance in the SDK and on the server.
         * The new EnumerationType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        EnumerationType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, EnumerationType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, EnumerationType, "javaType", false);
        };
        /**
         * Creates and returns a new EnumerationType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        EnumerationType.create = function (model) {
            return internal.instancehelpers.createElement(model, EnumerationType);
        };
        EnumerationType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return EnumerationType;
    }(Type));
    EnumerationType.typeName = "JavaActions$EnumerationType";
    EnumerationType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "properties": {
            "enumeration": {
                "public": {
                    "currentValue": true
                },
                "required": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.EnumerationType = EnumerationType;
    /**
     * In version 6.6.0: introduced
     */
    var FloatType = (function (_super) {
        __extends(FloatType, _super);
        function FloatType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new FloatType() cannot be invoked directly, please use 'model.javaactions.createFloatType()'");
            }
            return _this;
        }
        Object.defineProperty(FloatType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FloatType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FloatType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new FloatType instance in the SDK and on the server.
         * The new FloatType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        FloatType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, FloatType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, FloatType, "type", false);
        };
        /**
         * Creates and returns a new FloatType instance in the SDK and on the server.
         * The new FloatType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        FloatType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, FloatType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, FloatType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new FloatType instance in the SDK and on the server.
         * The new FloatType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        FloatType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, FloatType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, FloatType, "javaType", false);
        };
        /**
         * Creates and returns a new FloatType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        FloatType.create = function (model) {
            return internal.instancehelpers.createElement(model, FloatType);
        };
        FloatType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return FloatType;
    }(PrimitiveType));
    FloatType.typeName = "JavaActions$FloatType";
    FloatType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.FloatType = FloatType;
    /**
     * In version 6.6.0: introduced
     */
    var IntegerType = (function (_super) {
        __extends(IntegerType, _super);
        function IntegerType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new IntegerType() cannot be invoked directly, please use 'model.javaactions.createIntegerType()'");
            }
            return _this;
        }
        Object.defineProperty(IntegerType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(IntegerType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(IntegerType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new IntegerType instance in the SDK and on the server.
         * The new IntegerType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        IntegerType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, IntegerType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, IntegerType, "type", false);
        };
        /**
         * Creates and returns a new IntegerType instance in the SDK and on the server.
         * The new IntegerType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        IntegerType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, IntegerType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, IntegerType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new IntegerType instance in the SDK and on the server.
         * The new IntegerType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        IntegerType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, IntegerType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, IntegerType, "javaType", false);
        };
        /**
         * Creates and returns a new IntegerType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        IntegerType.create = function (model) {
            return internal.instancehelpers.createElement(model, IntegerType);
        };
        IntegerType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return IntegerType;
    }(PrimitiveType));
    IntegerType.typeName = "JavaActions$IntegerType";
    IntegerType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.IntegerType = IntegerType;
    /**
     * See: {@link https://docs.mendix.com/refguide7/java-actions relevant section in reference guide}
     */
    var JavaAction = (function (_super) {
        __extends(JavaAction, _super);
        function JavaAction(model, typeName, id, isPartial, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, container) || this;
            _this.__typeParameters = new internal.PartListProperty(JavaAction, _this, "typeParameters", []);
            _this.__parameters = new internal.PartListProperty(JavaAction, _this, "parameters", []);
            _this.__returnType = new internal.PrimitiveProperty(JavaAction, _this, "returnType", "", internal.PrimitiveTypeEnum.String);
            _this.__javaReturnType = new internal.PartProperty(JavaAction, _this, "javaReturnType", null, true);
            _this.__microflowActionInfo = new internal.PartProperty(JavaAction, _this, "microflowActionInfo", null, false);
            _this._containmentName = "documents";
            return _this;
        }
        Object.defineProperty(JavaAction.prototype, "containerAsFolderBase", {
            get: function () { return _super.prototype.getContainerAs.call(this, projects_1.projects.FolderBase); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaAction.prototype, "typeParameters", {
            /**
             * In version 6.6.0: introduced
             */
            get: function () { return this.__typeParameters.get(); },
            set: function (newValue) { this.__typeParameters.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaAction.prototype, "parameters", {
            get: function () { return this.__parameters.get(); },
            set: function (newValue) { this.__parameters.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaAction.prototype, "returnType", {
            /**
             * The value of this property is conceptually of type dataTypes.DataType.
             *
             * In version 6.6.0: deleted
             */
            get: function () { return this.__returnType.get(); },
            set: function (newValue) { this.__returnType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaAction.prototype, "javaReturnType", {
            /**
             * In version 6.6.0: introduced
             */
            get: function () { return this.__javaReturnType.get(); },
            set: function (newValue) { this.__javaReturnType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaAction.prototype, "microflowActionInfo", {
            /**
             * In version 6.6.0: introduced
             */
            get: function () { return this.__microflowActionInfo.get(); },
            set: function (newValue) { this.__microflowActionInfo.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates a new JavaAction unit in the SDK and on the server.
         * Expects one argument, the projects.IFolderBase in which this unit is contained.
         */
        JavaAction.createIn = function (container) {
            return internal.instancehelpers.createUnit(container, JavaAction);
        };
        JavaAction.prototype._isByNameReferrable = function () { return true; };
        JavaAction.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            if (this.__javaReturnType.isAvailable) {
                this.javaReturnType = BooleanType.create(this.model);
            }
            if (this.__returnType.isAvailable) {
                this.returnType = "Boolean";
            }
        };
        return JavaAction;
    }(projects_1.projects.Document));
    JavaAction.typeName = "JavaActions$JavaAction";
    JavaAction.versionInfo = new exports.StructureVersionInfo({
        "properties": {
            "typeParameters": {
                "introduced": "6.6.0",
                "public": {
                    "currentValue": true
                }
            },
            "parameters": {
                "public": {
                    "currentValue": true
                }
            },
            "returnType": {
                "deleted": "6.6.0",
                "deletionMessage": null,
                "public": {
                    "currentValue": true
                }
            },
            "javaReturnType": {
                "introduced": "6.6.0",
                "public": {
                    "currentValue": true
                },
                "required": {
                    "currentValue": true
                }
            },
            "microflowActionInfo": {
                "introduced": "6.6.0",
                "public": {
                    "currentValue": true
                }
            }
        }
    }, internal.StructureType.ModelUnit);
    javaactions.JavaAction = JavaAction;
    var JavaActionParameter = (function (_super) {
        __extends(JavaActionParameter, _super);
        function JavaActionParameter(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__name = new internal.PrimitiveProperty(JavaActionParameter, _this, "name", "", internal.PrimitiveTypeEnum.String);
            _this.__type = new internal.PrimitiveProperty(JavaActionParameter, _this, "type", "", internal.PrimitiveTypeEnum.String);
            _this.__javaType = new internal.PartProperty(JavaActionParameter, _this, "javaType", null, true);
            _this.__parameterType = new internal.PartProperty(JavaActionParameter, _this, "parameterType", null, true);
            _this.__description = new internal.PrimitiveProperty(JavaActionParameter, _this, "description", "", internal.PrimitiveTypeEnum.String);
            if (arguments.length < 4) {
                throw new Error("new JavaActionParameter() cannot be invoked directly, please use 'model.javaactions.createJavaActionParameter()'");
            }
            return _this;
        }
        Object.defineProperty(JavaActionParameter.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaActionParameter.prototype, "name", {
            get: function () { return this.__name.get(); },
            set: function (newValue) { this.__name.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaActionParameter.prototype, "type", {
            /**
             * The value of this property is conceptually of type dataTypes.DataType.
             *
             * In version 6.6.0: deleted
             */
            get: function () { return this.__type.get(); },
            set: function (newValue) { this.__type.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaActionParameter.prototype, "javaType", {
            /**
             * In version 6.7.0: deleted
             * In version 6.6.0: introduced
             */
            get: function () { return this.__javaType.get(); },
            set: function (newValue) { this.__javaType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaActionParameter.prototype, "parameterType", {
            /**
             * In version 6.7.0: introduced
             */
            get: function () { return this.__parameterType.get(); },
            set: function (newValue) { this.__parameterType.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(JavaActionParameter.prototype, "description", {
            /**
             * In version 6.10.0: introduced
             */
            get: function () { return this.__description.get(); },
            set: function (newValue) { this.__description.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new JavaActionParameter instance in the SDK and on the server.
         * The new JavaActionParameter will be automatically stored in the 'parameters' property
         * of the parent JavaAction element passed as argument.
         */
        JavaActionParameter.createIn = function (container) {
            return internal.instancehelpers.createElement(container, JavaActionParameter, "parameters", true);
        };
        /**
         * Creates and returns a new JavaActionParameter instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        JavaActionParameter.create = function (model) {
            return internal.instancehelpers.createElement(model, JavaActionParameter);
        };
        JavaActionParameter.prototype._isByNameReferrable = function () { return true; };
        JavaActionParameter.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            if (this.__javaType.isAvailable) {
                this.javaType = StringType.create(this.model);
            }
            if (this.__parameterType.isAvailable) {
                this.parameterType = BasicParameterType.create(this.model);
            }
            if (this.__type.isAvailable) {
                this.type = "Unknown";
            }
        };
        return JavaActionParameter;
    }(internal.Element));
    JavaActionParameter.typeName = "JavaActions$JavaActionParameter";
    JavaActionParameter.versionInfo = new exports.StructureVersionInfo({
        "properties": {
            "name": {
                "public": {
                    "currentValue": true
                }
            },
            "type": {
                "deleted": "6.6.0",
                "deletionMessage": null,
                "public": {
                    "currentValue": true
                }
            },
            "javaType": {
                "introduced": "6.6.0",
                "deleted": "6.7.0",
                "deletionMessage": null,
                "public": {
                    "currentValue": true
                },
                "required": {
                    "currentValue": true
                }
            },
            "parameterType": {
                "introduced": "6.7.0",
                "public": {
                    "currentValue": true
                },
                "required": {
                    "currentValue": true
                }
            },
            "description": {
                "introduced": "6.10.0",
                "public": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.JavaActionParameter = JavaActionParameter;
    /**
     * In version 6.6.0: introduced
     */
    var ListType = (function (_super) {
        __extends(ListType, _super);
        function ListType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__parameter = new internal.PartProperty(ListType, _this, "parameter", null, true);
            if (arguments.length < 4) {
                throw new Error("new ListType() cannot be invoked directly, please use 'model.javaactions.createListType()'");
            }
            return _this;
        }
        Object.defineProperty(ListType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ListType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ListType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ListType.prototype, "parameter", {
            get: function () { return this.__parameter.get(); },
            set: function (newValue) { this.__parameter.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new ListType instance in the SDK and on the server.
         * The new ListType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        ListType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, ListType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, ListType, "type", false);
        };
        /**
         * Creates and returns a new ListType instance in the SDK and on the server.
         * The new ListType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        ListType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, ListType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, ListType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new ListType instance in the SDK and on the server.
         * The new ListType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        ListType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, ListType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, ListType, "javaType", false);
        };
        /**
         * Creates and returns a new ListType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        ListType.create = function (model) {
            return internal.instancehelpers.createElement(model, ListType);
        };
        ListType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.parameter = ParameterizedEntityType.create(this.model);
        };
        return ListType;
    }(Type));
    ListType.typeName = "JavaActions$ListType";
    ListType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "properties": {
            "parameter": {
                "public": {
                    "currentValue": true
                },
                "required": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.ListType = ListType;
    /**
     * In version 6.10.0: removed experimental
     * In version 6.6.0: introduced
     */
    var MicroflowActionInfo = (function (_super) {
        __extends(MicroflowActionInfo, _super);
        function MicroflowActionInfo(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__caption = new internal.PrimitiveProperty(MicroflowActionInfo, _this, "caption", "", internal.PrimitiveTypeEnum.String);
            _this.__category = new internal.PrimitiveProperty(MicroflowActionInfo, _this, "category", "", internal.PrimitiveTypeEnum.String);
            _this.__icon = new internal.ByNameReferenceProperty(MicroflowActionInfo, _this, "icon", null, "Images$Image");
            if (arguments.length < 4) {
                throw new Error("new MicroflowActionInfo() cannot be invoked directly, please use 'model.javaactions.createMicroflowActionInfo()'");
            }
            return _this;
        }
        Object.defineProperty(MicroflowActionInfo.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MicroflowActionInfo.prototype, "caption", {
            get: function () { return this.__caption.get(); },
            set: function (newValue) { this.__caption.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MicroflowActionInfo.prototype, "category", {
            get: function () { return this.__category.get(); },
            set: function (newValue) { this.__category.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MicroflowActionInfo.prototype, "icon", {
            get: function () { return this.__icon.get(); },
            set: function (newValue) { this.__icon.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MicroflowActionInfo.prototype, "iconQualifiedName", {
            get: function () { return this.__icon.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new MicroflowActionInfo instance in the SDK and on the server.
         * The new MicroflowActionInfo will be automatically stored in the 'microflowActionInfo' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        MicroflowActionInfo.createIn = function (container) {
            internal.createInVersionCheck(container.model, MicroflowActionInfo.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, MicroflowActionInfo, "microflowActionInfo", false);
        };
        /**
         * Creates and returns a new MicroflowActionInfo instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        MicroflowActionInfo.create = function (model) {
            return internal.instancehelpers.createElement(model, MicroflowActionInfo);
        };
        MicroflowActionInfo.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return MicroflowActionInfo;
    }(internal.Element));
    MicroflowActionInfo.typeName = "JavaActions$MicroflowActionInfo";
    MicroflowActionInfo.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "properties": {
            "caption": {
                "public": {
                    "currentValue": true
                }
            },
            "category": {
                "public": {
                    "currentValue": true
                }
            },
            "icon": {
                "public": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        },
        "experimental": {
            "currentValue": false,
            "changedIn": [
                "6.10.0"
            ]
        }
    }, internal.StructureType.Element);
    javaactions.MicroflowActionInfo = MicroflowActionInfo;
    /**
     * In version 6.9.0: introduced
     */
    var MicroflowParameterType = (function (_super) {
        __extends(MicroflowParameterType, _super);
        function MicroflowParameterType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new MicroflowParameterType() cannot be invoked directly, please use 'model.javaactions.createMicroflowParameterType()'");
            }
            return _this;
        }
        Object.defineProperty(MicroflowParameterType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new MicroflowParameterType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        MicroflowParameterType.create = function (model) {
            return internal.instancehelpers.createElement(model, MicroflowParameterType);
        };
        MicroflowParameterType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return MicroflowParameterType;
    }(ParameterType));
    MicroflowParameterType.typeName = "JavaActions$MicroflowParameterType";
    MicroflowParameterType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.9.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.MicroflowParameterType = MicroflowParameterType;
    /**
     * In version 6.6.0: introduced
     */
    var ParameterizedEntityType = (function (_super) {
        __extends(ParameterizedEntityType, _super);
        function ParameterizedEntityType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__typeParameter = new internal.ByIdReferenceProperty(ParameterizedEntityType, _this, "typeParameter", null);
            if (arguments.length < 4) {
                throw new Error("new ParameterizedEntityType() cannot be invoked directly, please use 'model.javaactions.createParameterizedEntityType()'");
            }
            return _this;
        }
        Object.defineProperty(ParameterizedEntityType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ParameterizedEntityType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ParameterizedEntityType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ParameterizedEntityType.prototype, "containerAsListType", {
            get: function () { return _super.prototype.getContainerAs.call(this, ListType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(ParameterizedEntityType.prototype, "typeParameter", {
            get: function () { return this.__typeParameter.get(); },
            set: function (newValue) { this.__typeParameter.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new ParameterizedEntityType instance in the SDK and on the server.
         * The new ParameterizedEntityType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        ParameterizedEntityType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, ParameterizedEntityType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, ParameterizedEntityType, "type", false);
        };
        /**
         * Creates and returns a new ParameterizedEntityType instance in the SDK and on the server.
         * The new ParameterizedEntityType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        ParameterizedEntityType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, ParameterizedEntityType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, ParameterizedEntityType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new ParameterizedEntityType instance in the SDK and on the server.
         * The new ParameterizedEntityType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        ParameterizedEntityType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, ParameterizedEntityType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, ParameterizedEntityType, "javaType", false);
        };
        /**
         * Creates and returns a new ParameterizedEntityType instance in the SDK and on the server.
         * The new ParameterizedEntityType will be automatically stored in the 'parameter' property
         * of the parent ListType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        ParameterizedEntityType.createInListTypeUnderParameter = function (container) {
            internal.createInVersionCheck(container.model, ParameterizedEntityType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, ParameterizedEntityType, "parameter", false);
        };
        /**
         * Creates and returns a new ParameterizedEntityType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        ParameterizedEntityType.create = function (model) {
            return internal.instancehelpers.createElement(model, ParameterizedEntityType);
        };
        ParameterizedEntityType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return ParameterizedEntityType;
    }(EntityType));
    ParameterizedEntityType.typeName = "JavaActions$ParameterizedEntityType";
    ParameterizedEntityType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "properties": {
            "typeParameter": {
                "public": {
                    "currentValue": true
                },
                "required": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.ParameterizedEntityType = ParameterizedEntityType;
    /**
     * In version 6.6.0: introduced
     */
    var StringType = (function (_super) {
        __extends(StringType, _super);
        function StringType(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new StringType() cannot be invoked directly, please use 'model.javaactions.createStringType()'");
            }
            return _this;
        }
        Object.defineProperty(StringType.prototype, "containerAsBasicParameterType", {
            get: function () { return _super.prototype.getContainerAs.call(this, BasicParameterType); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(StringType.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(StringType.prototype, "containerAsJavaActionParameter", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaActionParameter); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new StringType instance in the SDK and on the server.
         * The new StringType will be automatically stored in the 'type' property
         * of the parent BasicParameterType element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        StringType.createInBasicParameterTypeUnderType = function (container) {
            internal.createInVersionCheck(container.model, StringType.typeName, ["6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, StringType, "type", false);
        };
        /**
         * Creates and returns a new StringType instance in the SDK and on the server.
         * The new StringType will be automatically stored in the 'javaReturnType' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        StringType.createInJavaActionUnderJavaReturnType = function (container) {
            internal.createInVersionCheck(container.model, StringType.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, StringType, "javaReturnType", false);
        };
        /**
         * Creates and returns a new StringType instance in the SDK and on the server.
         * The new StringType will be automatically stored in the 'javaType' property
         * of the parent JavaActionParameter element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0
         */
        StringType.createInJavaActionParameterUnderJavaType = function (container) {
            internal.createInVersionCheck(container.model, StringType.typeName, ["6.6.0"]);
            return internal.instancehelpers.createElement(container, StringType, "javaType", false);
        };
        /**
         * Creates and returns a new StringType instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        StringType.create = function (model) {
            return internal.instancehelpers.createElement(model, StringType);
        };
        StringType.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return StringType;
    }(PrimitiveType));
    StringType.typeName = "JavaActions$StringType";
    StringType.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.StringType = StringType;
    /**
     * In version 6.6.0: introduced
     */
    var TypeParameter = (function (_super) {
        __extends(TypeParameter, _super);
        function TypeParameter(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__name = new internal.PrimitiveProperty(TypeParameter, _this, "name", "", internal.PrimitiveTypeEnum.String);
            if (arguments.length < 4) {
                throw new Error("new TypeParameter() cannot be invoked directly, please use 'model.javaactions.createTypeParameter()'");
            }
            return _this;
        }
        Object.defineProperty(TypeParameter.prototype, "containerAsJavaAction", {
            get: function () { return _super.prototype.getContainerAs.call(this, JavaAction); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TypeParameter.prototype, "name", {
            get: function () { return this.__name.get(); },
            set: function (newValue) { this.__name.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new TypeParameter instance in the SDK and on the server.
         * The new TypeParameter will be automatically stored in the 'typeParameters' property
         * of the parent JavaAction element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        TypeParameter.createIn = function (container) {
            internal.createInVersionCheck(container.model, TypeParameter.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, TypeParameter, "typeParameters", true);
        };
        /**
         * Creates and returns a new TypeParameter instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        TypeParameter.create = function (model) {
            return internal.instancehelpers.createElement(model, TypeParameter);
        };
        TypeParameter.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return TypeParameter;
    }(internal.Element));
    TypeParameter.typeName = "JavaActions$TypeParameter";
    TypeParameter.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0",
        "properties": {
            "name": {
                "public": {
                    "currentValue": true
                }
            }
        },
        "public": {
            "currentValue": true
        }
    }, internal.StructureType.Element);
    javaactions.TypeParameter = TypeParameter;
})(javaactions = exports.javaactions || (exports.javaactions = {}));
