// mendixmodelsdk/dist/gen/navigation.js
/* tslint:disable */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var internal = require("../sdk/internal");
exports.StructureVersionInfo = internal.StructureVersionInfo;
var projects_1 = require("./projects");
var navigation;
(function (navigation) {
    var DeviceType = (function (_super) {
        __extends(DeviceType, _super);
        function DeviceType() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.qualifiedTsTypeName = "navigation.DeviceType";
            return _this;
        }
        return DeviceType;
    }(internal.IEnum));
    DeviceType.Desktop = new DeviceType("Desktop", {});
    DeviceType.Tablet = new DeviceType("Tablet", {});
    DeviceType.Phone = new DeviceType("Phone", {});
    navigation.DeviceType = DeviceType;
    var ProfileType = (function (_super) {
        __extends(ProfileType, _super);
        function ProfileType() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.qualifiedTsTypeName = "navigation.ProfileType";
            return _this;
        }
        return ProfileType;
    }(internal.IEnum));
    ProfileType.Desktop = new ProfileType("Desktop", {});
    ProfileType.Tablet = new ProfileType("Tablet", {});
    ProfileType.Phone = new ProfileType("Phone", {});
    ProfileType.OfflinePhone = new ProfileType("OfflinePhone", {});
    navigation.ProfileType = ProfileType;
    /**
     * Warning: this class represents an abstract and non-instantiable concept in the meta model.
     * You should _not_ try to instantiate this class directly, but instantiate one of its concrete/non-abstract sub classes instead.
     * Instantiating this class is possible in pure JavaScript, but there's no guarantee the result will be at all useful.
     */
    var HomePageBase = (function (_super) {
        __extends(HomePageBase, _super);
        function HomePageBase(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__page = new internal.ByNameReferenceProperty(HomePageBase, _this, "page", null, "Pages$Page");
            _this.__microflow = new internal.ByNameReferenceProperty(HomePageBase, _this, "microflow", null, "Microflows$Microflow");
            if (arguments.length < 4) {
                throw new Error("new HomePageBase() cannot be invoked directly, please use 'model.navigation.createHomePageBase()'");
            }
            return _this;
        }
        Object.defineProperty(HomePageBase.prototype, "containerAsNavigationProfile", {
            get: function () { return _super.prototype.getContainerAs.call(this, NavigationProfile); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(HomePageBase.prototype, "page", {
            get: function () { return this.__page.get(); },
            set: function (newValue) { this.__page.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(HomePageBase.prototype, "pageQualifiedName", {
            get: function () { return this.__page.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(HomePageBase.prototype, "microflow", {
            get: function () { return this.__microflow.get(); },
            set: function (newValue) { this.__microflow.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(HomePageBase.prototype, "microflowQualifiedName", {
            get: function () { return this.__microflow.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        HomePageBase.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return HomePageBase;
    }(internal.Element));
    HomePageBase.typeName = "Navigation$HomePageBase";
    HomePageBase.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    navigation.HomePageBase = HomePageBase;
    var HomePage = (function (_super) {
        __extends(HomePage, _super);
        function HomePage(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new HomePage() cannot be invoked directly, please use 'model.navigation.createHomePage()'");
            }
            return _this;
        }
        Object.defineProperty(HomePage.prototype, "containerAsNavigationProfile", {
            get: function () { return _super.prototype.getContainerAs.call(this, NavigationProfile); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new HomePage instance in the SDK and on the server.
         * The new HomePage will be automatically stored in the 'homePage' property
         * of the parent NavigationProfile element passed as argument.
         */
        HomePage.createIn = function (container) {
            return internal.instancehelpers.createElement(container, HomePage, "homePage", false);
        };
        /**
         * Creates and returns a new HomePage instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        HomePage.create = function (model) {
            return internal.instancehelpers.createElement(model, HomePage);
        };
        HomePage.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return HomePage;
    }(HomePageBase));
    HomePage.typeName = "Navigation$HomePage";
    HomePage.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    navigation.HomePage = HomePage;
    /**
     * See: {@link https://docs.mendix.com/refguide7/navigation relevant section in reference guide}
     */
    var NavigationDocument = (function (_super) {
        __extends(NavigationDocument, _super);
        function NavigationDocument(model, typeName, id, isPartial, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, container) || this;
            _this.__phoneProfile = new internal.PartProperty(NavigationDocument, _this, "phoneProfile", null, true);
            _this.__tabletProfile = new internal.PartProperty(NavigationDocument, _this, "tabletProfile", null, true);
            _this.__desktopProfile = new internal.PartProperty(NavigationDocument, _this, "desktopProfile", null, true);
            _this.__offlinePhoneProfile = new internal.PartProperty(NavigationDocument, _this, "offlinePhoneProfile", null, true);
            _this._containmentName = "projectDocuments";
            return _this;
        }
        Object.defineProperty(NavigationDocument.prototype, "containerAsProject", {
            get: function () { return _super.prototype.getContainerAs.call(this, projects_1.projects.Project); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationDocument.prototype, "phoneProfile", {
            get: function () { return this.__phoneProfile.get(); },
            set: function (newValue) { this.__phoneProfile.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationDocument.prototype, "tabletProfile", {
            get: function () { return this.__tabletProfile.get(); },
            set: function (newValue) { this.__tabletProfile.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationDocument.prototype, "desktopProfile", {
            get: function () { return this.__desktopProfile.get(); },
            set: function (newValue) { this.__desktopProfile.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationDocument.prototype, "offlinePhoneProfile", {
            get: function () { return this.__offlinePhoneProfile.get(); },
            set: function (newValue) { this.__offlinePhoneProfile.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates a new NavigationDocument unit in the SDK and on the server.
         * Expects one argument, the projects.IProject in which this unit is contained.
         */
        NavigationDocument.createIn = function (container) {
            return internal.instancehelpers.createUnit(container, NavigationDocument);
        };
        NavigationDocument.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
            this.desktopProfile = NavigationProfile.create(this.model);
            this.offlinePhoneProfile = NavigationProfile.create(this.model);
            this.phoneProfile = NavigationProfile.create(this.model);
            this.tabletProfile = NavigationProfile.create(this.model);
        };
        return NavigationDocument;
    }(projects_1.projects.ProjectDocument));
    NavigationDocument.typeName = "Navigation$NavigationDocument";
    NavigationDocument.versionInfo = new exports.StructureVersionInfo({
        "properties": {
            "phoneProfile": {
                "required": {
                    "currentValue": true
                }
            },
            "tabletProfile": {
                "required": {
                    "currentValue": true
                }
            },
            "desktopProfile": {
                "required": {
                    "currentValue": true
                }
            },
            "offlinePhoneProfile": {
                "required": {
                    "currentValue": true
                }
            }
        }
    }, internal.StructureType.ModelUnit);
    navigation.NavigationDocument = NavigationDocument;
    var NavigationProfile = (function (_super) {
        __extends(NavigationProfile, _super);
        function NavigationProfile(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__enabled = new internal.PrimitiveProperty(NavigationProfile, _this, "enabled", false, internal.PrimitiveTypeEnum.Boolean);
            _this.__homePage = new internal.PartProperty(NavigationProfile, _this, "homePage", null, true);
            _this.__roleBasedHomePages = new internal.PartListProperty(NavigationProfile, _this, "roleBasedHomePages", []);
            _this.__applicationTitle = new internal.PrimitiveProperty(NavigationProfile, _this, "applicationTitle", "", internal.PrimitiveTypeEnum.String);
            _this.__menuItemCollection = new internal.PartProperty(NavigationProfile, _this, "menuItemCollection", null, true);
            if (arguments.length < 4) {
                throw new Error("new NavigationProfile() cannot be invoked directly, please use 'model.navigation.createNavigationProfile()'");
            }
            return _this;
        }
        Object.defineProperty(NavigationProfile.prototype, "containerAsNavigationDocument", {
            get: function () { return _super.prototype.getContainerAs.call(this, NavigationDocument); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationProfile.prototype, "enabled", {
            get: function () { return this.__enabled.get(); },
            set: function (newValue) { this.__enabled.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationProfile.prototype, "homePage", {
            get: function () { return this.__homePage.get(); },
            set: function (newValue) { this.__homePage.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationProfile.prototype, "roleBasedHomePages", {
            get: function () { return this.__roleBasedHomePages.get(); },
            set: function (newValue) { this.__roleBasedHomePages.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationProfile.prototype, "applicationTitle", {
            get: function () { return this.__applicationTitle.get(); },
            set: function (newValue) { this.__applicationTitle.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NavigationProfile.prototype, "menuItemCollection", {
            get: function () { return this.__menuItemCollection.get(); },
            set: function (newValue) { this.__menuItemCollection.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new NavigationProfile instance in the SDK and on the server.
         * The new NavigationProfile will be automatically stored in the 'desktopProfile' property
         * of the parent NavigationDocument element passed as argument.
         */
        NavigationProfile.createInNavigationDocumentUnderDesktopProfile = function (container) {
            return internal.instancehelpers.createElement(container, NavigationProfile, "desktopProfile", false);
        };
        /**
         * Creates and returns a new NavigationProfile instance in the SDK and on the server.
         * The new NavigationProfile will be automatically stored in the 'offlinePhoneProfile' property
         * of the parent NavigationDocument element passed as argument.
         */
        NavigationProfile.createInNavigationDocumentUnderOfflinePhoneProfile = function (container) {
            return internal.instancehelpers.createElement(container, NavigationProfile, "offlinePhoneProfile", false);
        };
        /**
         * Creates and returns a new NavigationProfile instance in the SDK and on the server.
         * The new NavigationProfile will be automatically stored in the 'phoneProfile' property
         * of the parent NavigationDocument element passed as argument.
         */
        NavigationProfile.createInNavigationDocumentUnderPhoneProfile = function (container) {
            return internal.instancehelpers.createElement(container, NavigationProfile, "phoneProfile", false);
        };
        /**
         * Creates and returns a new NavigationProfile instance in the SDK and on the server.
         * The new NavigationProfile will be automatically stored in the 'tabletProfile' property
         * of the parent NavigationDocument element passed as argument.
         */
        NavigationProfile.createInNavigationDocumentUnderTabletProfile = function (container) {
            return internal.instancehelpers.createElement(container, NavigationProfile, "tabletProfile", false);
        };
        /**
         * Creates and returns a new NavigationProfile instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        NavigationProfile.create = function (model) {
            return internal.instancehelpers.createElement(model, NavigationProfile);
        };
        NavigationProfile.prototype._initializeDefaultProperties = function () {
            var _this = this;
            _super.prototype._initializeDefaultProperties.call(this);
            this.applicationTitle = (function () {
                if (internal.isNotBefore("[6.9.0]", _this.model)) {
                    return "Mendix";
                }
                return "Mendix 5";
            })();
            this.enabled = (function () {
                if (internal.isNotBefore("[6.9.0]", _this.model)) {
                    return false;
                }
                return true;
            })();
            this.homePage = HomePage.create(this.model);
            this.menuItemCollection = menus_1.menus.MenuItemCollection.create(this.model);
        };
        return NavigationProfile;
    }(internal.Element));
    NavigationProfile.typeName = "Navigation$NavigationProfile";
    NavigationProfile.versionInfo = new exports.StructureVersionInfo({
        "properties": {
            "enabled": {},
            "homePage": {
                "required": {
                    "currentValue": true
                }
            },
            "applicationTitle": {},
            "menuItemCollection": {
                "required": {
                    "currentValue": true
                }
            }
        }
    }, internal.StructureType.Element);
    navigation.NavigationProfile = NavigationProfile;
    var RoleBasedHomePage = (function (_super) {
        __extends(RoleBasedHomePage, _super);
        function RoleBasedHomePage(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__userRole = new internal.ByNameReferenceProperty(RoleBasedHomePage, _this, "userRole", null, "Security$UserRole");
            if (arguments.length < 4) {
                throw new Error("new RoleBasedHomePage() cannot be invoked directly, please use 'model.navigation.createRoleBasedHomePage()'");
            }
            return _this;
        }
        Object.defineProperty(RoleBasedHomePage.prototype, "containerAsNavigationProfile", {
            get: function () { return _super.prototype.getContainerAs.call(this, NavigationProfile); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(RoleBasedHomePage.prototype, "userRole", {
            get: function () { return this.__userRole.get(); },
            set: function (newValue) { this.__userRole.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(RoleBasedHomePage.prototype, "userRoleQualifiedName", {
            get: function () { return this.__userRole.qualifiedName(); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new RoleBasedHomePage instance in the SDK and on the server.
         * The new RoleBasedHomePage will be automatically stored in the 'roleBasedHomePages' property
         * of the parent NavigationProfile element passed as argument.
         */
        RoleBasedHomePage.createIn = function (container) {
            return internal.instancehelpers.createElement(container, RoleBasedHomePage, "roleBasedHomePages", true);
        };
        /**
         * Creates and returns a new RoleBasedHomePage instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        RoleBasedHomePage.create = function (model) {
            return internal.instancehelpers.createElement(model, RoleBasedHomePage);
        };
        RoleBasedHomePage.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return RoleBasedHomePage;
    }(HomePageBase));
    RoleBasedHomePage.typeName = "Navigation$RoleBasedHomePage";
    RoleBasedHomePage.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    navigation.RoleBasedHomePage = RoleBasedHomePage;
})(navigation = exports.navigation || (exports.navigation = {}));
var menus_1 = require("./menus");
