// mendixmodelsdk/dist/gen/xmlschemas.js
/* tslint:disable */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var internal = require("../sdk/internal");
exports.StructureVersionInfo = internal.StructureVersionInfo;
var projects_1 = require("./projects");
var mappings_1 = require("./mappings");
var xmlschemas;
(function (xmlschemas) {
    var XmlPrimitiveType = (function (_super) {
        __extends(XmlPrimitiveType, _super);
        function XmlPrimitiveType() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.qualifiedTsTypeName = "xmlschemas.XmlPrimitiveType";
            return _this;
        }
        return XmlPrimitiveType;
    }(internal.IEnum));
    XmlPrimitiveType.Unknown = new XmlPrimitiveType("Unknown", {});
    XmlPrimitiveType.Boolean = new XmlPrimitiveType("Boolean", {});
    XmlPrimitiveType.Date = new XmlPrimitiveType("Date", {});
    XmlPrimitiveType.Time = new XmlPrimitiveType("Time", {});
    XmlPrimitiveType.DateTime = new XmlPrimitiveType("DateTime", {});
    XmlPrimitiveType.Decimal = new XmlPrimitiveType("Decimal", {});
    XmlPrimitiveType.Float = new XmlPrimitiveType("Float", {});
    XmlPrimitiveType.Integer = new XmlPrimitiveType("Integer", {});
    XmlPrimitiveType.Long = new XmlPrimitiveType("Long", {
        "introduced": "6.2.0"
    });
    XmlPrimitiveType.Binary = new XmlPrimitiveType("Binary", {});
    XmlPrimitiveType.String = new XmlPrimitiveType("String", {});
    XmlPrimitiveType.AnyType = new XmlPrimitiveType("AnyType", {});
    xmlschemas.XmlPrimitiveType = XmlPrimitiveType;
    /**
     * Warning: this class represents an abstract and non-instantiable concept in the meta model.
     * You should _not_ try to instantiate this class directly, but instantiate one of its concrete/non-abstract sub classes instead.
     * Instantiating this class is possible in pure JavaScript, but there's no guarantee the result will be at all useful.
     */
    var MxSchema = (function (_super) {
        __extends(MxSchema, _super);
        function MxSchema(model, typeName, id, isPartial, container) {
            return _super.call(this, model, typeName, id, isPartial, container) || this;
        }
        Object.defineProperty(MxSchema.prototype, "containerAsFolderBase", {
            get: function () { return _super.prototype.getContainerAs.call(this, projects_1.projects.FolderBase); },
            enumerable: true,
            configurable: true
        });
        MxSchema.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return MxSchema;
    }(projects_1.projects.Document));
    MxSchema.typeName = "XmlSchemas$MxSchema";
    MxSchema.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.ModelUnit);
    xmlschemas.MxSchema = MxSchema;
    /**
     * In version 6.6.0: introduced
     */
    var XmlElement = (function (_super) {
        __extends(XmlElement, _super);
        function XmlElement(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            if (arguments.length < 4) {
                throw new Error("new XmlElement() cannot be invoked directly, please use 'model.xmlschemas.createXmlElement()'");
            }
            return _this;
        }
        Object.defineProperty(XmlElement.prototype, "containerAsJsonStructure", {
            get: function () { return _super.prototype.getContainerAs.call(this, jsonstructures_1.jsonstructures.JsonStructure); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlElement.prototype, "containerAsElement", {
            get: function () { return _super.prototype.getContainerAs.call(this, mappings_1.mappings.Element); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new XmlElement instance in the SDK and on the server.
         * The new XmlElement will be automatically stored in the 'elements' property
         * of the parent jsonstructures.JsonStructure element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        XmlElement.createInJsonStructureUnderElements = function (container) {
            internal.createInVersionCheck(container.model, XmlElement.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, XmlElement, "elements", true);
        };
        /**
         * Creates and returns a new XmlElement instance in the SDK and on the server.
         * The new XmlElement will be automatically stored in the 'children' property
         * of the parent mappings.Element element passed as argument.
         *
         * Warning! Can only be used on models with the following Mendix meta model versions:
         *  6.6.0, 6.7.0, 6.8.0, 6.9.0, 6.10.0, 7.0.0, 7.0.1
         */
        XmlElement.createInElementUnderChildren = function (container) {
            internal.createInVersionCheck(container.model, XmlElement.typeName, ["6.6.0", "6.7.0", "6.8.0", "6.9.0", "6.10.0", "7.0.0", "7.0.1"]);
            return internal.instancehelpers.createElement(container, XmlElement, "children", true);
        };
        /**
         * Creates and returns a new XmlElement instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        XmlElement.create = function (model) {
            return internal.instancehelpers.createElement(model, XmlElement);
        };
        XmlElement.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return XmlElement;
    }(mappings_1.mappings.Element));
    XmlElement.typeName = "XmlSchemas$XmlElement";
    XmlElement.versionInfo = new exports.StructureVersionInfo({
        "introduced": "6.6.0"
    }, internal.StructureType.Element);
    xmlschemas.XmlElement = XmlElement;
    /**
     * See: {@link https://docs.mendix.com/refguide7/xml-schemas relevant section in reference guide}
     */
    var XmlSchema = (function (_super) {
        __extends(XmlSchema, _super);
        function XmlSchema(model, typeName, id, isPartial, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, container) || this;
            _this.__entries = new internal.PartListProperty(XmlSchema, _this, "entries", []);
            _this.__filePath = new internal.PrimitiveProperty(XmlSchema, _this, "filePath", "", internal.PrimitiveTypeEnum.String);
            _this._containmentName = "documents";
            return _this;
        }
        Object.defineProperty(XmlSchema.prototype, "containerAsFolderBase", {
            get: function () { return _super.prototype.getContainerAs.call(this, projects_1.projects.FolderBase); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlSchema.prototype, "entries", {
            get: function () { return this.__entries.get(); },
            set: function (newValue) { this.__entries.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlSchema.prototype, "filePath", {
            get: function () { return this.__filePath.get(); },
            set: function (newValue) { this.__filePath.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates a new XmlSchema unit in the SDK and on the server.
         * Expects one argument, the projects.IFolderBase in which this unit is contained.
         */
        XmlSchema.createIn = function (container) {
            return internal.instancehelpers.createUnit(container, XmlSchema);
        };
        XmlSchema.prototype._isByNameReferrable = function () { return true; };
        XmlSchema.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return XmlSchema;
    }(MxSchema));
    XmlSchema.typeName = "XmlSchemas$XmlSchema";
    XmlSchema.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.ModelUnit);
    xmlschemas.XmlSchema = XmlSchema;
    var XmlSchemaEntry = (function (_super) {
        __extends(XmlSchemaEntry, _super);
        function XmlSchemaEntry(model, typeName, id, isPartial, unit, container) {
            var _this = _super.call(this, model, typeName, id, isPartial, unit, container) || this;
            _this.__targetNamespace = new internal.PrimitiveProperty(XmlSchemaEntry, _this, "targetNamespace", "", internal.PrimitiveTypeEnum.String);
            _this.__location = new internal.PrimitiveProperty(XmlSchemaEntry, _this, "location", "", internal.PrimitiveTypeEnum.String);
            _this.__contents = new internal.PrimitiveProperty(XmlSchemaEntry, _this, "contents", "", internal.PrimitiveTypeEnum.String);
            _this.__localizedLocationFormat = new internal.PrimitiveProperty(XmlSchemaEntry, _this, "localizedLocationFormat", "", internal.PrimitiveTypeEnum.String);
            _this.__localizedContentsFormat = new internal.PrimitiveProperty(XmlSchemaEntry, _this, "localizedContentsFormat", "", internal.PrimitiveTypeEnum.String);
            if (arguments.length < 4) {
                throw new Error("new XmlSchemaEntry() cannot be invoked directly, please use 'model.xmlschemas.createXmlSchemaEntry()'");
            }
            return _this;
        }
        Object.defineProperty(XmlSchemaEntry.prototype, "containerAsWsdlDescription", {
            get: function () { return _super.prototype.getContainerAs.call(this, webservices_1.webservices.WsdlDescription); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlSchemaEntry.prototype, "containerAsXmlSchema", {
            get: function () { return _super.prototype.getContainerAs.call(this, XmlSchema); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlSchemaEntry.prototype, "targetNamespace", {
            get: function () { return this.__targetNamespace.get(); },
            set: function (newValue) { this.__targetNamespace.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlSchemaEntry.prototype, "location", {
            get: function () { return this.__location.get(); },
            set: function (newValue) { this.__location.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlSchemaEntry.prototype, "contents", {
            get: function () { return this.__contents.get(); },
            set: function (newValue) { this.__contents.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlSchemaEntry.prototype, "localizedLocationFormat", {
            get: function () { return this.__localizedLocationFormat.get(); },
            set: function (newValue) { this.__localizedLocationFormat.set(newValue); },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(XmlSchemaEntry.prototype, "localizedContentsFormat", {
            get: function () { return this.__localizedContentsFormat.get(); },
            set: function (newValue) { this.__localizedContentsFormat.set(newValue); },
            enumerable: true,
            configurable: true
        });
        /**
         * Creates and returns a new XmlSchemaEntry instance in the SDK and on the server.
         * The new XmlSchemaEntry will be automatically stored in the 'schemaEntries' property
         * of the parent webservices.WsdlDescription element passed as argument.
         */
        XmlSchemaEntry.createInWsdlDescriptionUnderSchemaEntries = function (container) {
            return internal.instancehelpers.createElement(container, XmlSchemaEntry, "schemaEntries", true);
        };
        /**
         * Creates and returns a new XmlSchemaEntry instance in the SDK and on the server.
         * The new XmlSchemaEntry will be automatically stored in the 'entries' property
         * of the parent XmlSchema element passed as argument.
         */
        XmlSchemaEntry.createInXmlSchemaUnderEntries = function (container) {
            return internal.instancehelpers.createElement(container, XmlSchemaEntry, "entries", true);
        };
        /**
         * Creates and returns a new XmlSchemaEntry instance in the SDK and on the server.
         * Expects one argument: the IModel object the instance will "live on".
         * After creation, assign or add this instance to a property that accepts this kind of objects.
         */
        XmlSchemaEntry.create = function (model) {
            return internal.instancehelpers.createElement(model, XmlSchemaEntry);
        };
        XmlSchemaEntry.prototype._initializeDefaultProperties = function () {
            _super.prototype._initializeDefaultProperties.call(this);
        };
        return XmlSchemaEntry;
    }(internal.Element));
    XmlSchemaEntry.typeName = "XmlSchemas$XmlSchemaEntry";
    XmlSchemaEntry.versionInfo = new exports.StructureVersionInfo({}, internal.StructureType.Element);
    xmlschemas.XmlSchemaEntry = XmlSchemaEntry;
})(xmlschemas = exports.xmlschemas || (exports.xmlschemas = {}));
var jsonstructures_1 = require("./jsonstructures");
var webservices_1 = require("./webservices");
