// mendixmodelsdk/src/test/test-model.ts
import * as classes from "./gen/all-model-classes";
import {configuration} from "../sdk/config";
import * as base from "./gen/base-model";
import {ModelSdkClientImpl} from "../model-sdk-client";

export type IModel = base.IModel;

/**
 * Implementation of {@link IModel}.
 */
export class TestModel extends base.BaseModel implements IModel {

	private allModelClasses: any;

	/**
	 * Returns the structural unit of type "Projects$Project" which is the project's root,
	 * cast to the appropriate type.
	 */
	public get root(): classes.testmodel.IProject {
		if (this._unitsByType["TestModel$Project"].length !== 1) {
			throw new Error("Illegal State: project root data not available");
		}
		return this._unitsByType["TestModel$Project"][0] as classes.testmodel.IProject;
	}

	public static createSdkClient(connectionConfig: configuration.ISdkConfig) {
		return new ModelSdkClientImpl<IModel, TestModel>(connectionConfig, TestModel);
	}

	_allModelClasses() {
		if (!this.allModelClasses) {
			this.allModelClasses = require("./gen/all-model-classes");
		}
		return this.allModelClasses;
	}

}

export type TestModelSdkClient = ModelSdkClientImpl<IModel, TestModel>;
