// mendixmodelsdk/src/gen/expressions-types.ts
/* tslint:disable */
import {ITyped} from '../sdk/checks/base-types';

export namespace expressionsTypes {

    export /* abstract */ interface MxAbstractFollow extends MxBuiltInCall {
        path: MxExpression;
    }

    export interface MxBinaryExpression extends MxExpression {
        left: MxExpression;
        operator: MxOperator;
        right: MxExpression;
    }

    export interface MxBooleanLiteralExpression extends MxPrimitiveLiteralExpression {
        value: boolean;
    }

    export /* abstract */ interface MxBuiltInCall extends MxFeatureCall {
        variable: MxBuiltInVariableDeclaration;
    }

    export interface MxBuiltInVariableDeclaration extends MxNamed {
    }

    export interface MxColorLiteralExpression extends MxPrimitiveLiteralExpression {
        blue: number;
        green: number;
        red: number;
    }

    export interface MxDoubleLiteralExpression extends MxPrimitiveLiteralExpression {
        value: number;
    }

    export /* abstract */ interface MxExpression extends ITyped {
        appliedFeature: MxFeatureCallWrapper;
    }

    export /* abstract */ interface MxFeatureCall extends ITyped {
    }

    export interface MxFeatureCallWrapper extends MxExpression {
        featureCall: MxFeatureCall;
    }

    export interface MxFollowExcluding extends MxAbstractFollow {
    }

    export interface MxFollowIncluding extends MxAbstractFollow {
    }

    export interface MxIntegerLiteralExpression extends MxPrimitiveLiteralExpression {
        value: number;
    }

    export interface MxIteratorCall extends MxFeatureCall {
        body: MxExpression;
        type: MxIteratorType;
        variable: MxIteratorVariableDeclaration;
    }

    export interface MxIteratorVariableDeclaration extends MxNamed {
    }

    export interface MxListLiteralExpression extends MxLiteralExpression {
        items: MxExpression[];
    }

    export /* abstract */ interface MxLiteralExpression extends MxExpression {
    }

    export interface MxMethodCall extends MxFeatureCall {
        arguments: MxExpression[];
        referredMethod: string;
    }

    export /* abstract */ interface MxNamed extends ITyped {
        name: string;
    }

    export interface MxPrefixExpression extends MxExpression {
        expression: MxExpression;
    }

    export interface MxPreviousExpression extends MxExpression {
        variable: MxVariableRefExpression;
    }

    export /* abstract */ interface MxPrimitiveLiteralExpression extends MxSingleLiteralExpression {
    }

    export interface MxPropertyAssignment extends ITyped {
        property: string;
        value: MxExpression;
    }

    export interface MxPropertyCall extends MxFeatureCall {
        referredProperty: string;
    }

    export interface MxQualifiedEnumLiteralExpression extends MxSingleLiteralExpression {
        enumeration: string;
        literal: string;
    }

    export /* abstract */ interface MxSingleLiteralExpression extends MxLiteralExpression {
    }

    export interface MxSizeLiteralExpression extends MxPrimitiveLiteralExpression {
        height: number;
        width: number;
    }

    export interface MxStringLiteralExpression extends MxPrimitiveLiteralExpression {
        value: string;
    }

    export interface MxStructureLiteralExpression extends MxSingleLiteralExpression {
        assignments: MxPropertyAssignment[];
        structure: string;
    }

    export interface MxSwitch extends MxBuiltInCall {
        cases: MxSwitchCase[];
    }

    export interface MxSwitchCase extends ITyped {
        condition: MxExpression;
        expectedType: MxTypeStructureLiteralExpression;
        value: MxExpression;
    }

    export interface MxThisExpression extends MxExpression {
    }

    export /* abstract */ interface MxTypeLiteralExpression extends MxSingleLiteralExpression {
    }

    export interface MxTypeStructureLiteralExpression extends MxTypeLiteralExpression {
        referredStructure: string;
    }

    export interface MxUndefinedExpression extends MxExpression {
    }

    export interface MxVariableRefExpression extends MxExpression {
        referredVariable: string;
    }

    export type MxCheckSeverity = "NONE" | "WARNING" | "ERROR" | "DEPRECATION" | "TODO";

    export type MxIteratorType = "NONE" | "MAP" | "FILTER" | "EVERY" | "SOME" | "FIND_FIRST";

    export type MxOperator = "NONE" | "EQUALS" | "UNEQUALS" | "GREATER" | "LESS" | "GREATER_OR_EQUALS" | "LESS_OR_EQUALS" | "AND" | "OR";

    export type MxPrimitiveTypeEnum = "Integer" | "String" | "Boolean" | "Double" | "DateTime" | "Guid" | "Point" | "Size" | "Color" | "Blob";

}
